In [1]:
import os
os.chdir(r"C:\Users\Himanshu Poddar\Desktop\Specialisation\Data Science IBM(Python)\Capstone Project\Capstone_final_project\Rest_name_with_links")

In [2]:
import pandas as pd
import numpy as np

# Checking if all of my dataset for each category has 15 columns

In [3]:
neighborhoods = os.listdir()
for x in neighborhoods:
    types_rest = os.listdir(x)
    for rest in types_rest:
        if('meta' not in rest.lower()):
            df = pd.read_csv(r"" + x + "\\" + rest, encoding = "ISO-8859-1")
            if(df.shape[1]!=15):
                print(rest)

So our data collection procedure was correct and there is no anamoly in the number of columns

In [155]:
# Joining all the dataset since the dataset for each neighborhood and each category was scraped separately
listed_in = {'buffet': 'buffet_restaurants',
             'cafe': 'cafe_restaurants',
             'delivery': 'delivery_restaurants',
             'desserts': 'desserts&bakes',
             'dine': 'dineout_restaurants',
             'nightlife': 'drinks&nightlife',
             'pubs': 'pubs&bars'}
column_names = ['URL', 'Address', 'Name','Online_order', 'Book_table', 'Rate', 'Votes', 'Phone', 'Location', 'Rest_type', 'Dish_liked',
                              'Cuisines', 'Approx_cost', 'Reviews_list', 'Menu_item','Listed_in(type)', 'Listed_in(city)'] 

# instantiate the dataframe
main_frame = pd.DataFrame(columns=column_names)
main_frame

Empty DataFrame
Columns: [URL, Address, Name, Online_order, Book_table, Rate, Votes, Phone, Location, Rest_type, Dish_liked, Cuisines, Approx_cost, Reviews_list, Menu_item, Listed_in(type), Listed_in(city)]
Index: []

In [156]:
neighborhoods = os.listdir()
for x in neighborhoods:
    types_rest = os.listdir(x)
    for rest in types_rest:
        if('meta' not in rest.lower()):
            df = pd.read_csv(r"" + x + "\\" + rest, encoding = "ISO-8859-1")
            for key, value in listed_in.items():
                if key in rest:
                    df['Listed_in(type)'] = listed_in[key]
                    df['Listed_in(city)'] = x
                    # print(listed_in[key], rest, x)
                    
            #concat dataframes
            main_frame = pd.concat([main_frame, df], ignore_index=True)



In [157]:
# Lets check the total number of restaurants in Bengaluru
main_frame.shape

(51832, 17)

In [185]:
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate Votes                             Phone  \
0          Yes        Yes  4.1/5   775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5   787                      080 41714161   
2          Yes         No  3.8/5   918                    +91 9663487993   
3           No         No  3.7/5    88                    +91 9620009302   
4           No         No  3.8/5   166  +91 8026612447\r\n+91 9901210005   

       Location            Rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Café, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines                     Approx_cost  \
0  North Indian, Mughlai, Chinese  ?800 for two people (approx.)    
1     Chinese, North Indian, Thai  ?800 for two people (approx.)    
2          Cafe, Mexican, Italian  ?800 for two people (approx.)    
3      South Indian, North Indian  ?300 for two people (approx.)    
4        North Indian, Rajasthani  ?600 for two people (approx.)    

                                        Reviews_list Menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

      Listed_in(type) Listed_in(city)  
0  buffet_restaurants    Banashankari  
1  buffet_restaurants    Banashankari  
2  buffet_restaurants    Banashankari  
3  buffet_restaurants    Banashankari  
4  buffet_restaurants    Banashankari

#### Lets now check the total number of missing values in each column

In [186]:
# We can say that the URL of each restaurant in the dataset is our primary key
# So dropping all the data that are duplicate based on URL column
main_frame = main_frame.drop_duplicates('URL')

In [187]:
main_frame.shape

(51717, 17)

So approx 115 values were duplicate in our dataset

# Cleaning the data column wise

### <center>Phase I of cleaning</center>

### Lets one by one check for each column, if there is any noise in the data and clean parallely if any


#### URL column

In [188]:
main_frame['URL'].isnull().sum()

0

No null values in our URL column, which was quite obvious since we scraped the data using the url itself

#### Address column

In [189]:
main_frame['Address'].isnull().sum()

0

Our Address column also does not have any missing values

#### Name column

In [190]:
main_frame['Name'].isnull().sum()

0

No null values in Name, don't you think this was also obvious, "A restaurant without a name". Oh I think I got a name for my restaurant

#### Online_order column

In [191]:
main_frame['Online_order'].unique()

array(['Yes', 'No'], dtype=object)

No missing values in our Online_order column, contains only Yes or No

#### Book_table column

In [46]:
main_frame['Book_table'].unique()

array(['Yes', 'No'], dtype=object)

No missing values in our Book_table column, contains only Yes or No

#### Rate column

In [192]:
main_frame['Rate'].unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', '-', '2.6/5', nan, '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '4.8/5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.2 /5', '4.4 /5',
       '4.5 /5', '4.0 /5', '3.7 /5', '3.4 /5', '3.8 /5', '3.9 /5',
       '4.1 /5', '3.6 /5', '3.3 /5', '3.5 /5', '3.2 /5', '3.1 /5',
       '4.6 /5', '4.3 /5', '4.9 /5', '2.8 /5', '2.5 /5', '2.9 /5',
       '3.0 /5', '4.8 /5', '2.7 /5', '2.6 /5', '2.3 /5', '4.7 /5',
       '2.4 /5', '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

So there are some missing values e.g. nan, -, lets encode them with some value or leave them blank

In [28]:
na_values = ['-', np.nan]

Lets replace those NaN by blank

In [34]:
main_frame['Rate'].replace(np.nan, '', regex=True, inplace=True)
main_frame['Rate'].replace('-', '', regex=True, inplace=True)
main_frame['Rate'].unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', '', '2.6/5', '3.4/5', '4.5/5',
       '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5', '4.8/5',
       '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.2 /5', '4.4 /5', '4.5 /5',
       '4.0 /5', '3.7 /5', '3.4 /5', '3.8 /5', '3.9 /5', '4.1 /5',
       '3.6 /5', '3.3 /5', '3.5 /5', '3.2 /5', '3.1 /5', '4.6 /5',
       '4.3 /5', '4.9 /5', '2.8 /5', '2.5 /5', '2.9 /5', '3.0 /5',
       '4.8 /5', '2.7 /5', '2.6 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

We have replaced all the nan and its equivalent by empty string or blanks which will now be our symbol for nans

In [36]:
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate  Votes                             Phone  \
0          Yes        Yes  4.1/5  775.0    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5  787.0                      080 41714161   
2          Yes         No  3.8/5  918.0                    +91 9663487993   
3           No         No  3.7/5   88.0                    +91 9620009302   
4           No         No  3.8/5  166.0  +91 8026612447\r\n+91 9901210005   

       Location             Rest_type  \
0  Banashankari         Casual Dining   
1  Banashankari         Casual Dining   
2  Banashankari  Café, Casual Dining   
3  Banashankari           Quick Bites   
4  Basavanagudi         Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines                     Approx_cost  \
0  North Indian, Mughlai, Chinese  ?800 for two people (approx.)    
1     Chinese, North Indian, Thai  ?800 for two people (approx.)    
2          Cafe, Mexican, Italian  ?800 for two people (approx.)    
3      South Indian, North Indian  ?300 for two people (approx.)    
4        North Indian, Rajasthani  ?600 for two people (approx.)    

                                        Reviews_list Menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

      Listed_in(type) Listed_in(city)  
0  buffet_restaurants    Banashankari  
1  buffet_restaurants    Banashankari  
2  buffet_restaurants    Banashankari  
3  buffet_restaurants    Banashankari  
4  buffet_restaurants    Banashankari

#### Votes column

In [90]:
main_frame['Votes'].dtype

dtype('O')

So Votes column id of type float, Lets replace all the nans by blank

In [91]:
main_frame['Votes'].isnull().sum()

9934

There are 9,934 null values in our column

In [92]:
main_frame['Votes'].replace(np.nan, '', regex=True, inplace=True)

In [93]:
main_frame[main_frame['Votes'].isnull()]

Empty DataFrame
Columns: [URL, Address, Name, Online_order, Book_table, Rate, Votes, Phone, Location, Rest_type, Dish_liked, Cuisines, Approx_cost, Reviews_list, Menu_item, Listed_in(type), Listed_in(city)]
Index: []

None!!! Perfect!!!

#### Phone column

In [95]:
# Lets check whether phone number contains any character other than numbers and '+'
main_frame[~main_frame['Phone'].str.contains('[0-9+]',  na=False)]

URL  \
121    https://www.zomato.com/bangalore/new-mangalore...   
234    https://www.zomato.com/bangalore/kadamba-class...   
300    https://www.zomato.com/bangalore/just-thindi-b...   
378    https://www.zomato.com/bangalore/dakshin-kitch...   
466    https://www.zomato.com/bangalore/chettys-corne...   
467    https://www.zomato.com/bangalore/kapoors-cafe-...   
681    https://www.zomato.com/bangalore/parama-ruchi-...   
683    https://www.zomato.com/bangalore/meat-and-eat-...   
693    https://www.zomato.com/bangalore/hanumanthanag...   
727    https://www.zomato.com/bangalore/new-mangalore...   
816    https://www.zomato.com/bangalore/dakshin-kitch...   
821    https://www.zomato.com/bangalore/kadamba-class...   
831    https://www.zomato.com/bangalore/tejas-bar-and...   
836    https://www.zomato.com/bangalore/50-50-eating-...   
891    https://www.zomato.com/bangalore/starbucks-ban...   
922    https://www.zomato.com/bangalore/eat-fit-btm-b...   
1037   https://www.zomato.com/bangalore/lassi-shop-jp...   
1339   https://www.zomato.com/bangalore/kolkata-famou...   
1658   https://www.zomato.com/zyksha-food-truck?conte...   
1662   https://www.zomato.com/bangalore/papa-khan-ras...   
1768   https://www.zomato.com/bangalore/innate-jp-nag...   
1795   https://www.zomato.com/bangalore/super-chefs-b...   
1877   https://www.zomato.com/bangalore/rock-on-rolls...   
1896   https://www.zomato.com/bangalore/bowring-kulfi...   
1921   https://www.zomato.com/bangalore/venkateshwara...   
1940   https://www.zomato.com/bangalore/eat-fit-btm-b...   
1990   https://www.zomato.com/bangalore/starbucks-ban...   
1992   https://www.zomato.com/bangalore/kabab-magic-b...   
2084   https://www.zomato.com/bangalore/kailash-parba...   
2100   https://www.zomato.com/bangalore/spar-refresh-...   
...                                                  ...   
50215  https://www.zomato.com/bangalore/cafe-xtasi-va...   
50235  https://www.zomato.com/bangalore/beijing-bites...   
50248  https://www.zomato.com/bangalore/yumlane-brook...   
50303  https://www.zomato.com/bangalore/chef-delicaci...   
50328  https://www.zomato.com/bangalore/chatpata-indi...   
50405  https://www.zomato.com/bangalore/house-of-cand...   
50480  https://www.zomato.com/bangalore/tasty-box-whi...   
50631  https://www.zomato.com/bangalore/mayas-kitchen...   
50637  https://www.zomato.com/bangalore/mk-dabbawala-...   
50645  https://www.zomato.com/bangalore/kitchen-onlin...   
50647  https://www.zomato.com/bangalore/mast-kalandar...   
50702  https://www.zomato.com/bangalore/karavalli-foo...   
50733  https://www.zomato.com/bangalore/hunger-genie-...   
50755  https://www.zomato.com/bangalore/watheen-nuts-...   
50764  https://www.zomato.com/bangalore/innate-whitef...   
50860  https://www.zomato.com/bangalore/toastys-white...   
50932  https://www.zomato.com/bangalore/eat-fit-white...   
50940  https://www.zomato.com/bangalore/cinnabon-whit...   
50958  https://www.zomato.com/bangalore/auntie-annes-...   
50959  https://www.zomato.com/bangalore/starbucks-whi...   
51004  https://www.zomato.com/bangalore/cafe-coffee-d...   
51173  https://www.zomato.com/bangalore/starbucks-whi...   
51233  https://www.zomato.com/bangalore/mayas-kitchen...   
51247  https://www.zomato.com/bangalore/mast-kalandar...   
51271  https://www.zomato.com/bangalore/karavalli-foo...   
51353  https://www.zomato.com/bangalore/chatpata-indi...   
51501  https://www.zomato.com/bangalore/toastys-white...   
51591  https://www.zomato.com/bangalore/donne-biryani...   
51641  https://www.zomato.com/bangalore/andhra-spices...   
51714  https://www.zomato.com/bangalore/plunge-sherat...   

                                                 Address  \
121    89, Near PES College, 22nd Main Road, Girinaga...   
234    Satellite Bus Stand, KSRTC, Mysore Road, Banga...   
300    27, Gandhi Bazaar Main Rd, Basavanagudi, Banga...   
378    Kabirmutt Road, Gavipuram Extension, Basavanag...   
466    126/3, Uttarahalli Main R

So there are 1,208 data that does not have a phone number. Lets replace them by blank.

In [97]:
main_frame[~main_frame['Phone'].str.contains('[0-9+]',  na=False)]['Phone'].unique()

array(['Not Available', '', nan], dtype=object)

In [104]:
main_frame['Phone'].replace(np.nan, '', regex=True, inplace=True)
main_frame['Phone'].replace('Not Available', '', regex=True, inplace=True)

In [105]:
main_frame[~main_frame['Phone'].str.contains('[0-9+]',  na=False)]['Phone'].unique()

array([''], dtype=object)

All the Phone numbers are replaced by blank...

#### Location column

In [106]:
main_frame['Location'].isnull().sum()

0

In [107]:
main_frame[main_frame['Location'].isnull()]

Empty DataFrame
Columns: [URL, Address, Name, Online_order, Book_table, Rate, Votes, Phone, Location, Rest_type, Dish_liked, Cuisines, Approx_cost, Reviews_list, Menu_item, Listed_in(type), Listed_in(city)]
Index: []

No NULL values in Location column, PERFECT!

#### Rest_type columnm

In [110]:
main_frame['Rest_type'].isnull().sum()

205

There are 205 NULL values, lets have a look at them

In [114]:
main_frame[main_frame['Rest_type'].isnull()].head()

URL  \
999   https://www.zomato.com/bangalore/mughlai-kitch...   
1300  https://www.zomato.com/bangalore/eat-joy-resta...   
1376  https://www.zomato.com/bangalore/kababish-rest...   
1565  https://www.zomato.com/bangalore/goli-vada-pav...   
1637  https://www.zomato.com/bangalore/ammis-biryani...   

                                                Address                  Name  \
999   77/2 BDA Road, Opposite RR Layout, BTM 6th Sta...       Mughlai Kitchen   
1300  21, Saifee Apartment, 5th Cross, 22nd Main, Ne...  Eat & Joy Restaurant   
1376           7th Main, 18th Cross NS Palya, 2nd Stage   Kababish Restaurant   
1565  81/3, Hulimavu Road, Opposite Meenakshi Mall, ...    Goli Vada Pav No 1   
1637  3, 1st Cross, Kalyani Kala Mandir, Bannerghatt...        Ammi's Biryani   

     Online_order Book_table   Rate Votes                               Phone  \
999           Yes         No  3.5/5   269  +91 9663197947\r\r\n+91 9986692090   
1300          Yes         No  3.0/5     7                      +91 9945052352   
1376          Yes         No  3.6/5    15      +91 8050732123\r\n00 805074123   
1565           No         No    NaN          +91 9900561532\r\n+91 8971051846   
1637           No         No  2.4/5   392      +91 9513777152\r\n080 39457777   

               Location Rest_type  \
999                 BTM       NaN   
1300           JP Nagar       NaN   
1376                BTM       NaN   
1565  Bannerghatta Road       NaN   
1637  Bannerghatta Road       NaN   

                                             Dish_liked  \
999                                   Momos, Oreo Shake   
1300                                                NaN   
1376                                                NaN   
1565                                                NaN   
1637  Chicken Biryani, Hyderabadi Biryani, Rolls, Mu...   

                                   Cuisines                     Approx_cost  \
999   Mughlai, North Indian, Chinese, Momos  ?600 for two people (approx.)    
1300                 Fast Food, Finger Food  ?300 for two people (approx.)    
1376         North Indian, Chinese, Arabian  ?700 for two people (approx.)    
1565                    Street Food, Burger  ?150 for two people (approx.)    
1637                  Biryani, North Indian  ?500 for two people (approx.)    

                                           Reviews_list  \
999   [('Rated 5.0', "RATED\n  Ordered Chicken Kadai...   
1300  [('Rated 1.0', 'RATED\n  I had ordered for one...   
1376  [('Rated 2.0', 'RATED\n  You would only go to ...   
1565                                                 []   
1637  [('Rated 3.0', 'RATED\n  If you a spicy biriya...   

                                              Menu_item       Listed_in(type)  \
999   ['Fry Chicken Kabab [5 Pieces]', 'Fry Chicken ...  delivery_restaurants   
1300                                                 []  delivery_restaurants   
1376                                                 []  delivery_restaurants   
1565                                                 []  delivery_restaurants   
1637                                                 []  delivery_restaurants   

        Listed_in(city)  
999   Bannerghatta Road  
1300  Bannerghatta Road  
1376  Bannerghatta Road  
1565  Bannerghatta Road  
1637  Bannerghatta Road

Lets replace them by blank

In [115]:
main_frame['Rest_type'].replace(np.nan, '', regex=True, inplace=True)

In [117]:
main_frame['Location'].isnull().sum()

0

#### Dish_liked column

In [122]:
print("NA Values : ",main_frame['Dish_liked'].isnull().sum())
print("Blank : ", (main_frame['Dish_liked']=="").sum())

NA Values :  27880
Blank :  198


27,880 NULL values and 198 blank values, lets replace all NULL by blank


In [123]:
main_frame['Dish_liked'].replace(np.nan, '', regex=True, inplace=True)

In [125]:
print("NA Values : ",main_frame['Dish_liked'].isnull().sum())

NA Values :  0


#### Cuisines column

In [130]:
print("NULL values : ",main_frame['Cuisines'].isnull().sum())
print("Blank : ", (main_frame['Cuisines']=="").sum())

NULL values :  24
Blank :  21


Lets replace all NULL values by blank

In [133]:
main_frame['Cuisines'].replace(np.nan, '', regex=True, inplace=True)

In [6]:
print("NULL values : ",main_frame['Cuisines'].isnull().sum())

NULL values :  0


#### Approx_cost column

In [11]:
# Lets check the number of null values in the column
print("NULL values : ",main_frame['Approx_cost'].isnull().sum())
print("Blank : ", (main_frame['Approx_cost']=="").sum())

NULL values :  346
Blank :  0


Check if All the rows of approx cost contains "two people" in it, if yes then we can retain only the price for this column

In [18]:
main_frame[~main_frame['Approx_cost'].str.contains('two people', na=False)].shape

(346, 17)

346 values, Perfect!!! Hence we can say that except the rows that contains NULL values all the rows have the word *two people* in it. Lets extract approx cost using regex

In [52]:
main_frame['Approx_cost'].str.extract(r'(\d?,?\d+)').isnull().sum()

0    346
dtype: int64

So regex gives us the right values, since it also found 346 rows not to have any cost, and we know that those 346 values are our null values.
Lets have a look at them.

In [53]:
main_frame[main_frame['Approx_cost'].str.extract(r'(\d?,?\d+)').isnull().iloc[:,0]].head()

URL  \
1662  https://www.zomato.com/bangalore/papa-khan-ras...   
1768  https://www.zomato.com/bangalore/innate-jp-nag...   
1795  https://www.zomato.com/bangalore/super-chefs-b...   
3069  https://www.zomato.com/bangalore/super-chefs-b...   
3774  https://www.zomato.com/bangalore/the-hub-ibis-...   

                                                Address  \
1662  6, Abbiah Layout KC Halli Main Road, Bommanaha...   
1768  45, 14th Main, 7th phase, Puttehahalli Palya, ...   
1795  8/3, Opposite Suryadhoya Convention hall, Bann...   
3069  8/3, Opposite Suryadhoya Convention hall, Bann...   
3774  Ibis Bengaluru, Opposite RMZ Ecospace Business...   

                                Name Online_order Book_table   Rate  Votes  \
1662                 Papa Khan Rasoi           No         No    NaN    NaN   
1768                          Innate           No         No    NaN    NaN   
1795  Super Chef's (New Royal treat)           No         No    NaN    NaN   
3069  Super Chef's (New Royal treat)           No         No      -    NaN   
3774        The Hub - Ibis Bengaluru           No         No  3.9/5  237.0   

     Phone           Location    Rest_type  \
1662   NaN                NaN          NaN   
1768   NaN           JP Nagar     Delivery   
1795   NaN  Bannerghatta Road  Quick Bites   
3069   NaN  Bannerghatta Road  Quick Bites   
3774   NaN       Marathahalli          Bar   

                                             Dish_liked  \
1662                                                NaN   
1768                                                NaN   
1795                                                NaN   
3069                                                NaN   
3774  Dal Makhani, Cocktails, Lunch Buffet, Breakfas...   

                                 Cuisines Approx_cost  \
1662                                  NaN         NaN   
1768                            Beverages         NaN   
1795  North Indian, Chinese, South Indian         NaN   
3069  North Indian, Chinese, South Indian         NaN   
3774                          Finger Food         NaN   

                                           Reviews_list Menu_item  \
1662                                                 []        []   
1768                                                 []        []   
1795  [('Rated 1.0', 'RATED\n  Many items were missi...        []   
3069  [('Rated 1.0', 'RATED\n  Many items were missi...        []   
3774  [('Rated 4.0', 'RATED\n  The Hub is a small ba...        []   

           Listed_in(type)    Listed_in(city)  
1662  delivery_restaurants  Bannerghatta Road  
1768  delivery_restaurants  Bannerghatta Road  
1795  delivery_restaurants  Bannerghatta Road  
3069  delivery_restaurants       Basavanagudi  
3774    buffet_restaurants          Bellandur

Lets retain only the price values and rename the column by Approx_cost(for two people)

In [59]:
main_frame['Approx_cost'] = main_frame['Approx_cost'].str.extract(r'(\d?,?\d+)')

In [62]:
main_frame.rename(index=str, columns={"Approx_cost": "Approx_cost(for two people)"}, inplace=True)
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate  Votes                             Phone  \
0          Yes        Yes  4.1/5  775.0    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5  787.0                      080 41714161   
2          Yes         No  3.8/5  918.0                    +91 9663487993   
3           No         No  3.7/5   88.0                    +91 9620009302   
4           No         No  3.8/5  166.0  +91 8026612447\r\n+91 9901210005   

       Location                   Rest_type  \
0  Banashankari               Casual Dining   
1  Banashankari               Casual Dining   
2  Banashankari  Café, Casual Dining   
3  Banashankari                 Quick Bites   
4  Basavanagudi               Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines Approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        Reviews_list Menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

      Listed_in(type) Listed_in(city)  
0  buffet_restaurants    Banashankari  
1  buffet_restaurants    Banashankari  
2  buffet_restaurants    Banashankari  
3  buffet_restaurants    Banashankari  
4  buffet_restaurants    Banashankari

#### Reviews_list column

In [148]:
# lets check the NULL values in Reviews_list column
print("NULL values : ",main_frame['Reviews_list'].isnull().sum())
print("Blank : ", (main_frame['Reviews_list']=="[]").sum())

NULL values :  0
Blank :  7595


7595 Review_list are empty

In [149]:
main_frame[main_frame['Reviews_list']=="[]"].head()

URL  \
75   https://www.zomato.com/bangalore/om-sri-vinaya...   
91   https://www.zomato.com/bangalore/sunsadm-banas...   
92   https://www.zomato.com/bangalore/annapooranesh...   
95   https://www.zomato.com/bangalore/vi-ras-bar-an...   
107  https://www.zomato.com/bangalore/coffee-shopee...   

                                               Address  \
75   39, S B M Colony, Near-Sita Circle, 1st Stage,...   
91   Shop 18 , 3rd cross road,Markandeya Layout, Ba...   
92   625, 5th Cross Road, Hanumanthnagar, Banashank...   
95   3828, Near Deve Gowda Petrol Bunk, 30th Cross,...   
107  1722, 13th Main Road 31st Cross BSK, 2nd Stage...   

                           Name Online_order Book_table   Rate  Votes  \
75        Om Sri Vinayaka Chats           No         No    NEW    NaN   
91                      Sunsadm           No         No    NaN    NaN   
92       Annapooraneshwari Mess           No         No    NaN    NaN   
95   Vi Ra's Bar and Restaurant           No         No  3.3/5   62.0   
107               Coffee Shopee           No         No    NaN    NaN   

                                Phone      Location           Rest_type  \
75                     +91 8553206035  Banashankari  Takeaway, Delivery   
91                     +91 7829290358  Banashankari  Takeaway, Delivery   
92                     +91 9742009518  Banashankari                Mess   
95   +91 9845063930\r\n+91 7411148111  Banashankari                 Bar   
107                    +91 9591504624  Banashankari  Takeaway, Delivery   

         Dish_liked                             Cuisines  \
75              NaN               Street Food, Fast Food   
91              NaN                         South Indian   
92              NaN                         South Indian   
95   Chilli Chicken  South Indian, North Indian, Chinese   
107             NaN                            Beverages   

    Approx_cost(for two people) Reviews_list Menu_item       Listed_in(type)  \
75                          500           []        []  delivery_restaurants   
91                          400           []        []  delivery_restaurants   
92                          200           []        []  delivery_restaurants   
95                          800           []        []  delivery_restaurants   
107                         250           []        []  delivery_restaurants   

    Listed_in(city)  
75     Banashankari  
91     Banashankari  
92     Banashankari  
95     Banashankari  
107    Banashankari

Reviews_list is the list of tuples. Each tuple consists of two values. One is the rating and the other is the review. Since after reading from the file they would have been converted back to string. We need to convert it back to list of tuple format

In [6]:
import ast
main_frame['Reviews_list'] =  main_frame['Reviews_list'].apply(ast.literal_eval)
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate  Votes                             Phone  \
0          Yes        Yes  4.1/5  775.0    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5  787.0                      080 41714161   
2          Yes         No  3.8/5  918.0                    +91 9663487993   
3           No         No  3.7/5   88.0                    +91 9620009302   
4           No         No  3.8/5  166.0  +91 8026612447\r\n+91 9901210005   

       Location                           Rest_type  \
0  Banashankari                       Casual Dining   
1  Banashankari                       Casual Dining   
2  Banashankari  Café, Casual Dining   
3  Banashankari                         Quick Bites   
4  Basavanagudi                       Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines Approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        Reviews_list Menu_item  \
0  [(Rated 4.0, RATED\n  A beautiful place to din...        []   
1  [(Rated 4.0, RATED\n  Had been here for dinner...        []   
2  [(Rated 3.0, RATED\n  Ambience is not that goo...        []   
3  [(Rated 4.0, RATED\n  Great food and proper Ka...        []   
4  [(Rated 4.0, RATED\n  Very good restaurant in ...        []   

      Listed_in(type) Listed_in(city)  
0  buffet_restaurants    Banashankari  
1  buffet_restaurants    Banashankari  
2  buffet_restaurants    Banashankari  
3  buffet_restaurants    Banashankari  
4  buffet_restaurants    Banashankari

Lets count the number of reviews in our dataset

In [14]:
main_frame['Reviews_list'].apply(len).sum()

1320051

We have 1,320,051 reviews in our datset

#### Menu_item column

In [11]:
# lets check the NULL values in Menu_item column
print("NULL values : ",main_frame['Menu_item'].isnull().sum())
print("Blank : ", (main_frame['Menu_item']=="[]").sum())

NULL values :  0
Blank :  39617


There are 39,617 Menu_item that are empty. The menu item is a list of menus which are right now in string format. Lets convert them back to list format

In [41]:
main_frame['Menu_item'] = main_frame['Menu_item'].apply(ast.literal_eval)
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate  Votes                             Phone  \
0          Yes        Yes  4.1/5  775.0    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5  787.0                      080 41714161   
2          Yes         No  3.8/5  918.0                    +91 9663487993   
3           No         No  3.7/5   88.0                    +91 9620009302   
4           No         No  3.8/5  166.0  +91 8026612447\r\n+91 9901210005   

       Location                           Rest_type  \
0  Banashankari                       Casual Dining   
1  Banashankari                       Casual Dining   
2  Banashankari  Café, Casual Dining   
3  Banashankari                         Quick Bites   
4  Basavanagudi                       Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines Approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        Reviews_list Menu_item  \
0  [(Rated 4.0, RATED\n  A beautiful place to din...        []   
1  [(Rated 4.0, RATED\n  Had been here for dinner...        []   
2  [(Rated 3.0, RATED\n  Ambience is not that goo...        []   
3  [(Rated 4.0, RATED\n  Great food and proper Ka...        []   
4  [(Rated 4.0, RATED\n  Very good restaurant in ...        []   

      Listed_in(type) Listed_in(city)  
0  buffet_restaurants    Banashankari  
1  buffet_restaurants    Banashankari  
2  buffet_restaurants    Banashankari  
3  buffet_restaurants    Banashankari  
4  buffet_restaurants    Banashankari

#### Listed_in(type) column

In [46]:
# lets check the NULL values in Listed_in(type) column
print("NULL values : ",main_frame['Listed_in(type)'].isnull().sum())
print("Blank : ", (main_frame['Listed_in(type)']=="").sum())


NULL values :  0
Blank :  0


Voila!!! None

#### Listed_in(city) column

In [48]:
print("NULL values : ",main_frame['Listed_in(city)'].isnull().sum())
print("Blank : ", (main_frame['Listed_in(city)']=="").sum())

NULL values :  0
Blank :  0


No Null or Blank

### <center>Phase II of cleaning</center>

On going through the data once we found no noise or anamoly in our URL, Address, Name, Online_order, Book_table, Rate, Votes column.

#### Phone column

In the Phone column we saw that if the restaurant has multiple phone number they are separated by \r\n or \n. Lets replace all those phone number that contains letters other than space character, 0-9, \r, \n, plus(+) by blank.

In [114]:
main_frame['Phone'] = main_frame['Phone'].str.replace('[^\+0-9\r\n\s]','')

Lets check if any column contains phone other than our permissible character.

In [119]:
main_frame['Phone'].str.contains('[^\+0-9\r\n\s]',na = False).sum()

0

None!!! Perfect

#### Rest_type column

In [147]:
main_frame['Rest_type'].unique()

array(['Casual Dining',
       'CafÃ\x83Â\x83Ã\x82Â\x83Ã\x83Â\x82Ã\x82Â\x83Ã\x83Â\x83Ã\x82Â\x82Ã\x83Â\x82Ã\x82©, Casual Dining',
       'Quick Bites',
       'Casual Dining, CafÃ\x83Â\x83Ã\x82Â\x83Ã\x83Â\x82Ã\x82Â\x83Ã\x83Â\x83Ã\x82Â\x82Ã\x83Â\x82Ã\x82©',
       'CafÃ\x83Â\x83Ã\x82Â\x83Ã\x83Â\x82Ã\x82Â\x83Ã\x83Â\x83Ã\x82Â\x82Ã\x83Â\x82Ã\x82©',
       'Quick Bites, CafÃ\x83Â\x83Ã\x82Â\x83Ã\x83Â\x82Ã\x82Â\x83Ã\x83Â\x83Ã\x82Â\x82Ã\x83Â\x82Ã\x82©',
       'CafÃ\x83Â\x83Ã\x82Â\x83Ã\x83Â\x82Ã\x82Â\x83Ã\x83Â\x83Ã\x82Â\x82Ã\x83Â\x82Ã\x82©, Quick Bites',
       'Delivery', 'Mess', 'Dessert Parlor', 'Bakery, Dessert Parlor',
       'Pub', 'Bakery', 'Takeaway, Delivery', 'Fine Dining',
       'Beverage Shop', 'Sweet Shop', 'Bar', 'Beverage Shop, Quick Bites',
       'Confectionery', 'Quick Bites, Beverage Shop',
       'Dessert Parlor, Sweet Shop', 'Bakery, Quick Bites',
       'Sweet Shop, Quick Bites', 'Kiosk', 'Food Truck',
       'Quick Bites, Dessert Parlor', 'Beverage Shop, Dessert Par

After going through our data we also found that café is the only word that does not have English alphabet in it. What we can see is café is represented by its unicode character. lets replace it by normal e

In [151]:
main_frame['Rest_type'] = main_frame['Rest_type'].str.replace('[^A-Za-z\s,]','').str.replace('Caf','Cafe')

In [152]:
main_frame['Rest_type'].unique()

array(['Casual Dining', 'Cafe, Casual Dining', 'Quick Bites',
       'Casual Dining, Cafe', 'Cafe', 'Quick Bites, Cafe',
       'Cafe, Quick Bites', 'Delivery', 'Mess', 'Dessert Parlor',
       'Bakery, Dessert Parlor', 'Pub', 'Bakery', 'Takeaway, Delivery',
       'Fine Dining', 'Beverage Shop', 'Sweet Shop', 'Bar',
       'Beverage Shop, Quick Bites', 'Confectionery',
       'Quick Bites, Beverage Shop', 'Dessert Parlor, Sweet Shop',
       'Bakery, Quick Bites', 'Sweet Shop, Quick Bites', 'Kiosk',
       'Food Truck', 'Quick Bites, Dessert Parlor',
       'Beverage Shop, Dessert Parlor', 'Takeaway', 'Pub, Casual Dining',
       'Casual Dining, Bar', 'Dessert Parlor, Beverage Shop',
       'Quick Bites, Bakery', 'Dessert Parlor, Quick Bites',
       'Microbrewery, Casual Dining', 'Lounge', 'Bar, Casual Dining',
       'Food Court', 'Cafe, Bakery', nan, 'Dhaba',
       'Quick Bites, Sweet Shop', 'Microbrewery',
       'Food Court, Quick Bites', 'Pub, Bar', 'Casual Dining, Pub',
      

###### Good, It looks so clean now.

In [154]:
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate  Votes                             Phone  \
0          Yes        Yes  4.1/5  775.0    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5  787.0                      080 41714161   
2          Yes         No  3.8/5  918.0                    +91 9663487993   
3           No         No  3.7/5   88.0                    +91 9620009302   
4           No         No  3.8/5  166.0  +91 8026612447\r\n+91 9901210005   

       Location            Rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines Approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        Reviews_list Menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

      Listed_in(type) Listed_in(city)  
0  buffet_restaurants    Banashankari  
1  buffet_restaurants    Banashankari  
2  buffet_restaurants    Banashankari  
3  buffet_restaurants    Banashankari  
4  buffet_restaurants    Banashankari

#### Location column

In [167]:
main_frame['Location'].unique()

array(['Banashankari', 'Basavanagudi', 'Mysore Road', 'Jayanagar',
       'Kumaraswamy Layout', 'Rajarajeshwari Nagar', 'Vijay Nagar',
       'Uttarahalli', 'JP Nagar', 'South Bangalore', 'City Market',
       'Nagarbhavi', 'Bannerghatta Road', 'BTM', 'Kanakapura Road',
       'Bommanahalli', nan, 'CV Raman Nagar', 'Electronic City', 'HSR',
       'Marathahalli', 'Sarjapur Road', 'Wilson Garden', 'Shanti Nagar',
       'Koramangala 5th Block', 'Koramangala 8th Block', 'Richmond Road',
       'Koramangala 7th Block', 'Jalahalli', 'Koramangala 4th Block',
       'Bellandur', 'Whitefield', 'East Bangalore', 'Old Airport Road',
       'Indiranagar', 'Koramangala 1st Block', 'Frazer Town', 'RT Nagar',
       'MG Road', 'Brigade Road', 'Lavelle Road', 'Church Street',
       'Ulsoor', 'Residency Road', 'Shivajinagar', 'Infantry Road',
       'St. Marks Road', 'Cunningham Road', 'Race Course Road',
       'Commercial Street', 'Vasanth Nagar', 'HBR Layout', 'Domlur',
       'Ejipura', 'Jeevan 

In [168]:
main_frame['Location'].nunique()

93

So there are 92 neighborhoods in Bengaluru, 93rd is nan

#### Listed_in(type) column

In [172]:
main_frame['Listed_in(type)'].unique()

array(['buffet_restaurants', 'cafe_restaurants', 'delivery_restaurants',
       'desserts&bakes', 'dineout_restaurants', 'drinks&nightlife',
       'pubs&bars'], dtype=object)

In [175]:
type_of_meal = {'delivery_restaurants' : 'Delivery',
                'dineout_restaurants' : 'Dine-out',
                'drinks&nightlife' : 'Drinks & nightlife',
                'pubs&bars' : 'Pubs and bars',
                'buffet_restaurants' : 'Buffet',
                'cafe_restaurants' : 'Cafes',
                'desserts&bakes' : 'Desserts' }

In [179]:
main_frame['Listed_in(type)'] = main_frame['Listed_in(type)'].map(type_of_meal)
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate  Votes                             Phone  \
0          Yes        Yes  4.1/5  775.0    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5  787.0                      080 41714161   
2          Yes         No  3.8/5  918.0                    +91 9663487993   
3           No         No  3.7/5   88.0                    +91 9620009302   
4           No         No  3.8/5  166.0  +91 8026612447\r\n+91 9901210005   

       Location            Rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines Approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        Reviews_list Menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  Listed_in(type) Listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

In [180]:
main_frame['Listed_in(type)'].unique()

array(['Buffet', 'Cafes', 'Delivery', 'Desserts', 'Dine-out',
       'Drinks & nightlife', 'Pubs and bars'], dtype=object)

Great!!! All the types were replaced successfully

#### Listed_in(city) column

In [183]:
main_frame['Listed_in(city)'].unique()

array(['Banashankari', 'Bannerghatta Road', 'Basavanagudi', 'Bellandur',
       'Brigade Road', 'Brookefield', 'BTM', 'Church Street',
       'Electronic City', 'Frazer Town', 'HSR', 'Indiranagar',
       'Jayanagar', 'JP Nagar', 'Kalyan Nagar', 'Kammanahalli',
       'Koramangala 4th Block', 'Koramangala 5th Block',
       'Koramangala 6th Block', 'Koramangala 7th Block', 'Lavelle Road',
       'Malleshwaram', 'Marathahalli', 'MG Road', 'New BEL Road',
       'Old Airport Road', 'Rajajinagar', 'Residency Road',
       'Sarjapur Road', 'Whitefield'], dtype=object)

In [184]:
main_frame['Listed_in(city)'].nunique()

30

Listed_in(city) contains the restaurants data listed by city in the zomato bangalore main page. Although there are'nt only above mentioned neighborhoods in bangalore. But the data in the website was listed by cities this way. There were 30 neighborhoods which were separately listed.

In [187]:
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate  Votes                             Phone  \
0          Yes        Yes  4.1/5  775.0    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5  787.0                      080 41714161   
2          Yes         No  3.8/5  918.0                    +91 9663487993   
3           No         No  3.7/5   88.0                    +91 9620009302   
4           No         No  3.8/5  166.0  +91 8026612447\r\n+91 9901210005   

       Location            Rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines Approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        Reviews_list Menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  Listed_in(type) Listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

### <center>Phase III of cleaning</center>

Assigning appropriate data type and column name to our data

In [188]:
# Verifying data types for each column
main_frame.dtypes

URL                             object
Address                         object
Name                            object
Online_order                    object
Book_table                      object
Rate                            object
Votes                          float64
Phone                           object
Location                        object
Rest_type                       object
Dish_liked                      object
Cuisines                        object
Approx_cost(for two people)     object
Reviews_list                    object
Menu_item                       object
Listed_in(type)                 object
Listed_in(city)                 object
dtype: object

For each of the column assignng their appropriate categories based on the analysis we have done so far.

In [201]:
main_frame['Online_order'] = main_frame['Online_order'].astype('category')
main_frame['Book_table'] = main_frame['Book_table'].astype('category')
main_frame['Location'] = main_frame['Location'].astype('category')
main_frame['Rest_type'] = main_frame['Rest_type'].astype('category')
main_frame['Listed_in(type)'] = main_frame['Listed_in(type)'].astype('category')
main_frame['Listed_in(city)'] = main_frame['Listed_in(city)'].astype('category')

In [202]:
main_frame.dtypes

URL                              object
Address                          object
Name                             object
Online_order                   category
Book_table                     category
Rate                             object
Votes                           float64
Phone                            object
Location                       category
Rest_type                      category
Dish_liked                       object
Cuisines                         object
Approx_cost(for two people)      object
Reviews_list                     object
Menu_item                        object
Listed_in(type)                category
Listed_in(city)                category
dtype: object

In [203]:
main_frame.head()

URL  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             Address                   Name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  Online_order Book_table   Rate  Votes                             Phone  \
0          Yes        Yes  4.1/5  775.0    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5  787.0                      080 41714161   
2          Yes         No  3.8/5  918.0                    +91 9663487993   
3           No         No  3.7/5   88.0                    +91 9620009302   
4           No         No  3.8/5  166.0  +91 8026612447\r\n+91 9901210005   

       Location            Rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          Dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         Cuisines Approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        Reviews_list Menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  Listed_in(type) Listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

#### Converting Votes column from float to integer

Votes are always in integer but in our dataframe it is showing as float. Lets convert it back to integer

In [205]:
main_frame['Votes'] = main_frame['Votes'].fillna(0)
main_frame['Votes'] = main_frame['Votes'].astype(int)

In [210]:
main_frame.describe(include = 'all')

URL        Address  \
count                                               51717          51717   
unique                                              51717          11495   
top     https://www.zomato.com/bangalore/fresh-presser...  Delivery Only   
freq                                                    1            128   
mean                                                  NaN            NaN   
std                                                   NaN            NaN   
min                                                   NaN            NaN   
25%                                                   NaN            NaN   
50%                                                   NaN            NaN   
75%                                                   NaN            NaN   
max                                                   NaN            NaN   

                   Name Online_order Book_table   Rate         Votes  \
count             51717        51717      51717  43942  51717.000000   
unique             8792            2          2     64           NaN   
top     Cafe Coffee Day          Yes         No    NEW           NaN   
freq                 96        30444      45268   2208           NaN   
mean                NaN          NaN        NaN    NaN    283.697527   
std                 NaN          NaN        NaN    NaN    803.838853   
min                 NaN          NaN        NaN    NaN      0.000000   
25%                 NaN          NaN        NaN    NaN      7.000000   
50%                 NaN          NaN        NaN    NaN     41.000000   
75%                 NaN          NaN        NaN    NaN    198.000000   
max                 NaN          NaN        NaN    NaN  16832.000000   

               Phone Location    Rest_type Dish_liked      Cuisines  \
count          50509    51696        51490      23639         51672   
unique         14926       93           93       5271          2723   
top     080 43334321      BTM  Quick Bites    Biryani  North Indian   
freq             216     5124        19132        182          2913   
mean             NaN      NaN          NaN        NaN           NaN   
std              NaN      NaN          NaN        NaN           NaN   
min              NaN      NaN          NaN        NaN           NaN   
25%              NaN      NaN          NaN        NaN           NaN   
50%              NaN      NaN          NaN        NaN           NaN   
75%              NaN      NaN          NaN        NaN           NaN   
max              NaN      NaN          NaN        NaN           NaN   

       Approx_cost(for two people) Reviews_list Menu_item Listed_in(type)  \
count                        51371        51717     51717           51717   
unique                          70        22513      9098               7   
top                            300           []        []        Delivery   
freq                          7576         7595     39617           25942   
mean                           NaN          NaN       NaN             NaN   
std                            NaN          NaN       NaN             NaN   
min                            NaN          NaN       NaN             NaN   
25%                            NaN          NaN       NaN             NaN   
50%                            NaN          NaN       NaN             NaN   
75%                            NaN          NaN       NaN             NaN   
max                            NaN          NaN       NaN             NaN   

       Listed_in(city)  
count            51717  
unique              30  
top                BTM  
freq              3279  
mean               NaN  
std                NaN  
min                NaN  
25%                NaN  
50%                NaN  
75%                NaN  
max                NaN

#### Convert each of the column name to lower case character

In [213]:
main_frame.columns

Index(['URL', 'Address', 'Name', 'Online_order', 'Book_table', 'Rate', 'Votes',
       'Phone', 'Location', 'Rest_type', 'Dish_liked', 'Cuisines',
       'Approx_cost(for two people)', 'Reviews_list', 'Menu_item',
       'Listed_in(type)', 'Listed_in(city)'],
      dtype='object')

In [214]:
main_frame.columns = map(str.lower, main_frame.columns)
main_frame.columns

Index(['url', 'address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'reviews_list', 'menu_item',
       'listed_in(type)', 'listed_in(city)'],
      dtype='object')

##### One thing to be worth noting is the same restaurant can be listed in different category with different URL. hence URL cannot be the primary key

### <center>*** That's All Folks ***</center>

<center><font size="4" color="F4C613">That was all about cleaning of our data. Meet you in <B>Analysis</B> Phase.</font></center>

Double-click __here__ to import/export as pickle and save the data

<!--
main_frame.to_csv(r"C:\Users\Himanshu Poddar\Desktop\zomato_analysis.csv", index=False)
main_frame = pd.read_csv(r"C:\Users\Himanshu Poddar\Desktop\zomato_analysis.csv", encoding="ISO-8859-1")
import pickle

# Store data (serialize)
with open('filename.pickle', 'wb') as handle:
    pickle.dump(main_frame, handle, protocol=pickle.HIGHEST_PROTOCOL)

# Load data (deserialize)
with open('filename.pickle', 'rb') as handle:
    unserialized_data = pickle.load(handle)

print(your_data == unserialized_data)

-->